packages/contracts-bedrock/src/dispute/AnchorStateRegistry.sol

```diff
@@ -23,105 +22,222 @@ import { ISuperchainConfig } from "interfaces/L1/ISuperchainConfig.sol";
-        // Grab the game and game data.
-        IFaultDisputeGame game = IFaultDisputeGame(msg.sender);
-        (GameType gameType, Claim rootClaim, bytes memory extraData) = game.gameData();
+    function pokeAnchorState(uint256 _candidateGameIndex) external {
+        IFaultDisputeGame _game = anchorGameCandidates[_candidateGameIndex];
+        _anchorGameCandidateIndex = _candidateGameIndex + 1;
+        _updateAnchorState(_game);
+    }
 
-        // Grab the verified address of the game based on the game data.
-        // slither-disable-next-line unused-return
-        (IDisputeGame factoryRegisteredGame,) =
-            DISPUTE_GAME_FACTORY.games({ _gameType: gameType, _rootClaim: rootClaim, _extraData: extraData });
+    function _updateAnchorState(IFaultDisputeGame _game) internal {
+        uint256 _anchorL2BlockNumber = _anchorGame.l2BlockNumber();
+        uint256 _gameL2BlockNumber = _game.l2BlockNumber();
+        if (!isGameValid(_game)) revert InvalidGame();
+        if (_gameL2BlockNumber <= _anchorL2BlockNumber) {
+            revert AnchorGameIsNewer(_anchorL2BlockNumber, _gameL2BlockNumber);
+        }
+        _updateAnchorStateWithValidNewerGame(_game);
+    }
 
-        // Must be a valid game.
-        if (address(factoryRegisteredGame) != address(game)) revert UnregisteredGame();
+    function _updateAnchorStateWithValidNewerGame(IFaultDisputeGame _game) internal {
+        _anchorGame = _game;
+        emit AnchorGameSet(_game);
+    }
 
-        // No need to update anything if the anchor state is already newer.
-        if (game.l2BlockNumber() <= anchors[gameType].l2BlockNumber) {
+    function _maybeRegisterAnchorGameCandidate() internal {
+        IFaultDisputeGame _game = IFaultDisputeGame(msg.sender);
+        // game must not be invalid
+        if (!isGameMaybeValid(_game)) return;
+        // if the game is older than the anchor game, we don't need it
+        if (_game.l2BlockNumber() < _anchorGame.l2BlockNumber()) {
             return;
         }
+        anchorGameCandidates.push(_game);
+    }
 
-        // Must be a game that resolved in favor of the state.
-        if (game.status() != GameStatus.DEFENDER_WINS) {
-            return;
+    /// @notice Callable by FaultDisputeGame contracts to update the anchor state.
+    function tryUpdateAnchorState() external {
+        _maybeRegisterAnchorGameCandidate();
+        uint256 _anchorGameBlockNumber = _anchorGame.l2BlockNumber();
+        uint256 _gasStart = gasleft();
+        // TODO: add padding to ensure we don't run out of gas
+        while (_gasStart - gasleft() < tryUpdateAnchorStateGas) {
+            // if there are no candidates to consider, break
+            if (_anchorGameCandidateIndex == anchorGameCandidates.length) break;
+            // If the game's l2BlockNumber is older than that of anchorGame, seek ahead.
+            IFaultDisputeGame _anchorCandidate = anchorGameCandidates[_anchorGameCandidateIndex];
```

Comment:
here and elsewhere I use `IFaultDisputeGame` because `IDisputeGame` doesn't have `l2BlockNumber()` on it. I should look into whether we can add `l2BlockNumber()` to `IDisputeGame`.